### ISSUE TITLE
Segmentation fault using P/Invoke on Raspberry Pi 3B / ARM32

### ISSUE THREAD

**Issue by tillig** [dotnet/coreclr#22384]
# Summary

I'm trying to use P/Invoke to control an LED strip attached to my Raspberry Pi GPIO pins. [The source for the native C library I'm working with is here.](https://github.com/jgarff/rpi_ws281x)

I've found that I get a segmentation fault when trying to use the native interop support that can be inconsistent. I noticed there are several other segmentation fault on ARM/Raspberry Pi open and I may be running into the same thing.

# Environment

- Raspberry Pi 3B
- Raspbian 9
- .NET Core SDK 2.2.103

# Reproduction

I've created a minimal repro [in this repository](https://github.com/tillig/Ws281xArmSegmentationFault). It includes a built version of the native library. In that repository there is a `logs` folder that includes logs from running `strace` and `gdb`. The error, when seen, consistently reports in GDB as being from `WKS::gc_heap::mark_object_simple(unsigned char**) ()` in `libcoreclr.so`. I've also got some notes I'm keeping in the [readme in that repo](https://github.com/tillig/Ws281xArmSegmentationFault/blob/master/README.md) as I troubleshoot.

I've not yet done a core dump to analyze the issue, however I have tried setting `COMPlus_HeapVerify=1` like `sudo COMPlus_HeapVerify=1 ./ConsoleDemo` on the published application [as noted in this issue](https://github.com/dotnet/coreclr/issues/18564#issuecomment-398857248) and I didn't see any additional output. I'm not sure if I was supposed to or if I wasn't properly setting the value.

Something I did note which was very, very odd: While trying to debug the issue, I was adding some `Console.WriteLine()` calls to see how far the app would get before faulting. At one point I actually got it to run _entirely without any faults_ just by adding enough `Console.WriteLine()` calls with no other changes. I commented out the calls, re-published, re-ran... and got the fault again. Put the `Console.WriteLine()` calls back... no fault. I left the `Console.WriteLine()` calls there, added some more operations to the lights (more P/Invoke ops) and got the fault again. I just thought it was really weird that I could consisently get past the fault just by adding `Console.WriteLine()`. Didn't make any sense.

I intend on getting a core dump and trying to debug through it to see if I can get more information, but I thought I'd raise it here in case someone had something... maybe simpler I could try. Or perhaps this is already being solved in some other issue?

**Comment by tillig** [comment]
I updated my issue as I realized that I was on an RPi 3B, not 3B+. Unsure if that makes a difference, but in the interest of precision... :)

**Comment by janvorli** [comment]
I'm going to take a look.

**Comment by janvorli** [comment]
As a first step, I can confirm I can repro the issue.

**Comment by janvorli** [comment]
Using the `COMPlus_HeapVerify=1`, I can see that the heap got corrupted between two GCs, which may point to a possible issue with the native library. E.g. using a pointer to managed object that is not pinned etc.
Unfortunately lldb doesn't fully work on ARM32 and so I cannot use SOS plugin to dump various managed objects etc. This problem with lldb has been bothering me for a long time and so I am starting to look into it. 
I'll continue debugging this issue in a week after I get back from my vacation.

**Comment by tillig** [comment]
In case it helps... the native library is [from this repository](https://github.com/jgarff/rpi_ws281x). I built it with the instructions they provided:

```bash
apt-get install scons
scons
```

I then built the .so file with:

```bash
gcc -shared -o ws2811.so *.o
```

I admit a lack of much recent experience with C/C++ lower-level debugging so working through with `lldb` or whatever was something I was going to have to research and figure out on the fly. If it doesn't totally work on ARM32 I'd likely have been out of my depth. However, if there is additional info I can provide or things I can help with on this end, I'm absolutely open to that. I just need maybe a nudge in the right direction. 😄

**Comment by 0xd4d** [comment]
```C#
            this._ws2811Handle = GCHandle.Alloc(this._ws2811, GCHandleType.Pinned);
```

This boxes the value type stored in `_ws2811`. The next couple of lines overwrite some of its fields but it doesn't update the boxed value type. You can change it to a class (with sequential layout, just like structs) or allocate native memory (and get rid of GCHandle).

```C#
            var initResult = NativeMethods.ws2811_init(ref this._ws2811);
```

You need to use a 'fixed' statement (or allocate native memory and pass in the pointer). There are a few more similar method calls.

**Comment by jkoritzinsky** [comment]
@janvorli any update on this issue?

**Comment by janvorli** [comment]
@jkoritzinsky there were some issues in the code that @0xd4d pointed out. The 2nd issue could clearly cause the memory corruption.

**Comment by tillig** [comment]
Again, I apologize that I'm a bit out of my depth on this and I very much appreciate your help and patience with me. I'm trying to learn, but sometimes it helps to get a little bit more specific, if you can.

> `this._ws2811Handle = GCHandle.Alloc(this._ws2811, GCHandleType.Pinned);`
> This boxes the value type stored in `_ws2811`.  ...You can change it to a class (with sequential layout, just like structs)...

OK, so if I switch `_ws2811` to be a class instead of a struct then I won't need the `GCHandle` calls. I will have to try this. The `_ws2811` object has a property inside that is also a struct; I'm guessing I need to switch that child property to be a class. (?)

>             `var initResult = NativeMethods.ws2811_init(ref this._ws2811);`
> You need to use a 'fixed' statement (or allocate native memory and pass in the pointer). There are a few more similar method calls.

I'm finding that `fixed` only works on some primitive types. What is supposed to be `fixed` in this note - the result or the `_ws2811` object? The return type from this init call is an integer.

**Comment by janvorli** [comment]
Let me explain the issue with `var initResult = NativeMethods.ws2811_init(ref this._ws2811)` in your code in detail. The _ws2811 is a struct and that means it is a value type. As such, it is stored inside the memory allocated for the Ws281xController instance (which is the "this" here). You are passing address of the _ws2811 in this instance to the native code. So when GC moves the Ws281xController instance to another memory location, the address of _ws2811 changes with it, but your native code uses the old address. And on that address, now a different object may live. So storing something in there corrupts it.

You were tryign to prevent moving of the _ws2811 using a pinned GCHandle. However, such handle can only be created for reference (non-value) objects. And thus when you try to create the handle for the _ws2811 which is a value object, the runtime creates a boxed copy of the ws2811_t value and the GC handle pins that one. So this is not what you wanted to do.

So what you can do is to change the ws2811_t to class. That way, it won't be allocated inside of the memory allocated for the Ws281xController, but it will be a separate object on the heap and the Ws281xController will contain just a reference to that object. So you can then create a pinned handle for the instance of the ws2811_t, which will ensure that it doesn't move during GC until you free that handle.

You'd also need to create an instance of the ws2811_t in the Ws281xController constructor, so instead of 
```C#
this._ws2811 = default(ws2811_t);
```
you'd use
```C#
this._ws2811 = new ws2811_t();
```

**Comment by tillig** [comment]
Thank you so much for the clarification! I will give this a shot.

**Comment by tillig** [comment]
I switched the `ws2811_t` from `struct` to `class` but unfortunately I'm now getting

```
Unhandled Exception: System.ArgumentException: Object contains non-primitive or non-blittable data.
   at System.Runtime.InteropServices.GCHandle.InternalAlloc(Object value, GCHandleType type)
   at System.Runtime.InteropServices.GCHandle.Alloc(Object value, GCHandleType type)
   at AddressableLed.Ws281xController..ctor(Settings settings)
   at ConsoleDemo.Program.Main() in ConsoleDemo\Program.cs:line 18
```

The line failing is the `this._ws2811Handle = GCHandle.Alloc(this._ws2811, GCHandleType.Pinned);` call.

I noticed @0xd4d mentioned I'd need some `fixed` items, but the compiler is telling me that `fixed` can only be a member of a `struct`. I tried switching the `ws2811_channel_t` from `struct` to `class` but that didn't fix the issue, I still get the exception.

Again, I apologize for being slow on the uptake here.

**Comment by jkoritzinsky** [comment]
@tillig I just copied the definitions of `ws2811_channel_t` and `ws_2811_t` from your repository into a test environment and I'm not getting the same exception as you on either .NET Core or .NET Framework, which is very odd.

**Comment by janvorli** [comment]
I have tried the same thing yesterday and I also didn't have this problem. @tillig have you really just changed the struct keyword to class without additional changes?

**Comment by tillig** [comment]
I pushed my changes to the repro repo, did a `dotnet publish -r linux-arm` on the project, and tried to run the result on Raspbian on an RPi 3. Let me try again, maybe something went amiss?

**Comment by tillig** [comment]
You're right, I must have had something cached. I did a full `git clean -dfx`, republished, re-pushed to the Pi and now I'm getting a whole different error. I think this last bit could get me over the hump:

The original `ws2811_t` structure has an array of channels:

```c++
typedef struct ws2811_t
{
    uint64_t render_wait_time;                   //< time in µs before the next render can run
    struct ws2811_device *device;                //< Private data for driver use
    const rpi_hw_t *rpi_hw;                      //< RPI Hardware Information
    uint32_t freq;                               //< Required output frequency
    int dmanum;                                  //< DMA number _not_ already in use
    ws2811_channel_t channel[RPI_PWM_CHANNELS];
} ws2811_t;

typedef struct ws2811_channel_t
{
    int gpionum;                                 //< GPIO Pin with PWM alternate function, 0 if unused
    int invert;                                  //< Invert output signal
    int count;                                   //< Number of LEDs, 0 if channel is unused
    int strip_type;                              //< Strip color layout -- one of WS2811_STRIP_xxx constants
    ws2811_led_t *leds;                          //< LED buffers, allocated by driver based on count
    uint8_t brightness;                          //< Brightness value between 0 and 255
    uint8_t wshift;                              //< White shift value
    uint8_t rshift;                              //< Red shift value
    uint8_t gshift;                              //< Green shift value
    uint8_t bshift;                              //< Blue shift value
    uint8_t *gamma;                              //< Gamma correction table
} ws2811_channel_t;
```

I found elsewhere that marshaling an array with a fixed number of elements could be done by appending the index to the field name, which felt wrong but I tried it.

```c#
    [StructLayout(LayoutKind.Sequential)]
    internal class ws2811_t
    {
        public long render_wait_time;
        public IntPtr device;
        public IntPtr rpi_hw;
        public uint freq;
        public int dmanum;
        public ws2811_channel_t channel_1;
    }

    [StructLayout(LayoutKind.Sequential)]
    internal struct ws2811_channel_t
    {
        public int gpionum;
        public int invert;
        public int count;
        public int strip_type;
        public IntPtr leds;
        public byte brightness;
        public byte wshift;
        public byte rshift;
        public byte gshift;
        public byte bshift;
        public IntPtr gamma;
    }
```

The error I'm getting now is around that `ws2811_channel_t` object having an incorrect value in it, which indicates to me it may not be getting passed across properly.

I set the `gpionum` to 18 and the error says, "Gpio 2126597824 is illegal for LED channel 0" - not really close to 18. I feel like this indicates the data isn't getting marshaled properly. Is there a better way to handle that array?

(I recognize this is pretty much a new question. If you'd rather close this issue, I totally understand. In any case, I _very, very much appreciate_ the help that's been provided here so far!)

**Comment by janvorli** [comment]
As for the array of the channels, I've looked at the definition of the RPI_PWM_CHANNELS and it is 2. So the easiest way to fix that is to add one more member, `public ws2811_channel_t channel_2;`. There is a way to create an array that's inlined in a data structure (https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/unsafe-code-pointers/fixed-size-buffers), but it can be used only in structs and the element type can only be one of the basic numeric types (byte, short, int, ...). It would be doable to transform your code that way, but it doesn't seem worth all the added complexity.
The issue "Gpio 2126597824 is illegal for LED channel 0" is caused by incorrect marshalling of the ws_2811_t class. I can see that the native code is getting the address of the managed object and not the address of the first field. Each reference object (class) starts with a pointer to so called MethodTable, which describes the type of the object and the actual fields are stored after that. So the native code was getting address of that.
I am no expert on the interop, but what would definitely help is to change all the parameters of the methods in AddressableLed.Interop.NativeMethods from `ref ws2811_t ws2811` to `IntPtr ws2811` and then at all places where you call them pass in`this._ws2811Handle.AddrOfPinnedObject()` instead of `ref this._ws2811`). 
I've tried to do that and your stuff now seems to run. I cannot verify that it does what it is supposed to do, but it doesn't report any error and doesn't seem to crash.

**Comment by tillig** [comment]
**YES YES YES YES YES** You folks are my heroes. 🥇 🎉 

I had to:

- Switch the `NativeMethods` to use `IntPtr` as you mentioned; _and_
- Add the second channel (the `channel_2` structure) to match the setup in the main library (without this I was getting another segmentation fault)

But after that _it totally worked_. I quite literally ran around my living room pumping my fists in triumph.

Thank you ever so much for your help, it means the world.
